pkg/client/add_events.go

```diff
@@ -40,6 +42,79 @@ import (
 Wrapper around: https://app.scalyr.com/help/api#addEvents
 */
 
+type EventWithMeta struct {
+	EventBundle *add_events.EventBundle
+	Key         string
+	SessionInfo add_events.SessionInfo
+}
+
+func NewEventWithMeta(
+	bundle *add_events.EventBundle,
+	groupBy []string,
+	serverHost string,
+) EventWithMeta {
+	// initialise
+	key := ""
+	info := make(add_events.SessionInfo)
+
+	// figure out serverHost value
+	usedServerHost := serverHost
+	// if event's ServerHost is set, use it
+	if len(bundle.Event.ServerHost) > 0 {
+		usedServerHost = bundle.Event.ServerHost
+	} else {
+		// if somebody is using library directly and forget to set Event.ServerHost,
+		// lets check the attributes first
+
+		// check serverHost attribute
+		attrHost, ok := bundle.Event.Attrs[add_events.AttrServerHost]
+		if ok {
+			usedServerHost = attrHost.(string)
+		} else {
+			// if serverHost attribute is not set, check the __origServerHost
+			attrOrigHost, okOrig := bundle.Event.Attrs[add_events.AttrOrigServerHost]
+			if okOrig {
+				usedServerHost = attrOrigHost.(string)
+			}
+		}
+	}
+
+	// for the SessionInfo, it has to be in the serverHost
+	// therefore remove the orig and set the serverHost
+	// so that it can be in the next step moved to sessionInfo
+	delete(bundle.Event.Attrs, add_events.AttrOrigServerHost)
+	bundle.Event.Attrs[add_events.AttrServerHost] = usedServerHost
+
+	// construct Key
+	// move attributes from attrs to sessionInfo
+	for _, k := range groupBy {
```

Comment:
To make unit testing easier, should this piece of logic and one for setting the server host live in separate utility function?

Reply:
EDIT: I see you actually move it from a standalone function in line here. I personally think standalone function may make it a bit easier to read and directly test those standalone functions as well.

Reply:
I will move it. Initially there was more interactions between those two sections, but then I have made them more independent.

Reply:
Thanks.